src/parser/shared/modules/AbilityTracker.ts

```diff
@@ -0,0 +1,130 @@
+import RESOURCE_TYPES from 'game/RESOURCE_TYPES';
+import Analyzer from 'parser/core/Analyzer';
+import { Ability, AbsorbedEvent, CastEvent, DamageEvent, EventType, HealEvent } from 'parser/core/Events';
+import HIT_TYPES from 'game/HIT_TYPES';
+
+import SpellManaCost from './SpellManaCost';
+
+export interface TrackedAbility {
+  ability: Ability | null;
+  casts: number;
+  manaUsed: number;
+  damageHits: number;
+  damageEffective: number;
+  damageAbsorbed: number;
+  damageCriticalHits: number;
+  damageCriticalEffective: number;
+  damageCriticalAbsorbed: number;
+  healingHits: number;
+  healingEffective: number;
+  healingAbsorbed: number;
+  healingOverheal: number;
+  healingCriticalHits: number;
+  healingCriticalEffective: number;
+  healingCriticalAbsorbed: number;
+  healingCriticalOverheal: number;
+}
+
+class AbilityTracker extends Analyzer {
+  static dependencies = {
+    // Needed for the `resourceCost` prop of events
+    spellManaCost: SpellManaCost,
+  };
+  protected spellManaCost!: SpellManaCost;
+
+  abilities = new Map<number, TrackedAbility>();
+
+  on_byPlayer_cast(event: CastEvent) {
+    const spellId = event.ability.guid;
+
+    const cast = this.getAbility(spellId, event.ability);
+    cast.casts = cast.casts + 1;
+    if (event.resourceCost && event.resourceCost[RESOURCE_TYPES.MANA.id] !== undefined) {
+      cast.manaUsed = cast.manaUsed + event.resourceCost[RESOURCE_TYPES.MANA.id];
+    }
+  }
+
+  getAbility(spellId: number, abilityInfo: Ability | null = null) {
+    let ability = this.abilities.get(spellId);
+    if (ability === undefined) {
+      ability = {
```

Comment:
I believe its pretty safe to default all of these numerical values to zero, and it saves throughout the codebase the potential of them being undefined. Is there any cases people can think of where having one of these values set to undefined is useful to know, compared to it being zero?